drv/stm32h7-spi-rot-server/src/main.rs

```diff
@@ -0,0 +1,274 @@
+        // If SP and RoT are out of sync, e.g. this task restarts and an old
+        // response is still in the RoT's transmit FIFO, then we can also see
+        // ROT_IRQ asserted when not expected.
+
+        ringbuf_entry!(Trace::GpioPort(port, ROT_IRQ.pin_mask));
+        // Async and unhandled responses from RoT are not expected.
+        panic!();
+    }
+    if let Err(spi_error) = server.spi.write(&server.message[0..xmit_len]) {
+        ringbuf_entry!(Trace::SpiError(spi_error));
+        return Err(MsgError::SpiServerError);
+    }
+
+    // We sleep and poll for what should be long enough for the RoT
+    // to queue a response.
+    // TODO: For better performance and power efficiency,
+    // take an interrupt on ROT_IRQ falling edge with timeout.
+    const LIMIT_START: u64 = 1_000_000; // TODO: Something more reasonable.
+    const NAP: u64 = 1;
+    let mut limit = LIMIT_START; // 1.04ms Time limit is somewhat arbitrary.
+                                 // TODO: put this time limit in StatusRsp
+    loop {
+        if 0 == server.sys.gpio_read(ROT_IRQ).unwrap_lite() {
+            ringbuf_entry!(Trace::RotIrqAsserted);
+            break;
+        }
+        limit -= 1;
+        if limit == 0 {
+            ringbuf_entry!(Trace::RotIrqTimeout);
+            break;
+        }
+        hl::sleep_for(NAP);
+    }
+
+    /*
+    // Unmask our interrupt.
+
+    // STM32 EXTI allows for 16 interrupts for GPIOs.
+    // Each of those can represent Pin X from a GPIO bank (A through K)
+    // So, only one bank's Pin 3, for example, can have the #3 interrupt.
+    // For ROT_IRQ, we would configure for the falling edge to trigger
+    // the interrupt. That configuration should be specified in the app.toml
+    // for the board. Work needs to be done to generalize the EXTI facility.
+    // But, hacking in one interrupt as an example should be ok to start things
+    // off.
+
+    sys_irq_control(self.interrupt, true);
+    // And wait for it to arrive.
+    let _rm =
+        sys_recv_closed(&mut [], self.interrupt, TaskId::KERNEL)
+            .unwrap_lite();
+    */
+
+    // Read just the header.
+    // Keep CSn asserted over the two reads.
+    server
+        .spi
+        .lock(CsState::Asserted)
+        .map_err(|_| MsgError::SpiServerError)?;
+    ringbuf_entry!(Trace::Line);
+    if let Err(spi_error) =
+        server.spi.read(&mut server.message[0..SPI_HEADER_SIZE])
+    {
+        ringbuf_entry!(Trace::SpiError(spi_error));
+        server.spi.release().unwrap_lite();
+        return Err(MsgError::SpiServerError); // XXX spi_error value?
+    }
+
+    ringbuf_entry!(Trace::Data(
+        server.message[0],
+        server.message[1],
+        server.message[2],
+        server.message[3]
+    ));
+
+    let msg = Msg::parse(&mut server.message[..]).unwrap_lite();
+    if !msg.is_supported_version() {
+        ringbuf_entry!(Trace::BadProtocol(server.message[0]));
+        server.spi.release().unwrap_lite();
```

Comment:
What happens on the client side here? Does it end up confused if CS becomes deasserted midway through a transmission?

Reply:
The CSn assertion bounds the SPI frame. If the message doesn't validate, there will be an error response.